src/tracer/MARBL_tracers.F90

```diff
@@ -0,0 +1,734 @@
+  type(MOM_restart_CS), pointer :: restart_CSp => NULL() !< A pointer to the restart control structure
+
+  type(vardesc), allocatable :: tr_desc(:) !< Descriptions and metadata for the tracers
+  logical :: tracers_may_reinit = .false. !< If true the tracers may be initialized if not found in a restart file
+
+  !> Driver-specific parameters
+  character(len=35) :: marbl_settings_file
+  real :: atm_co2_const, atm_alt_co2_const
+
+  !> Indices to the registered diagnostics match the indices used in MARBL
+  type(temp_MARBL_diag), allocatable :: surface_flux_diags(:) !, interior_tendency_diags(:)
+  type(saved_state_for_MARBL_type), allocatable :: surface_flux_saved_state(:) !, interior_tendency_saved_state(:)
+  integer :: u10_sqr_ind, sss_ind, sst_ind, ifrac_ind, dust_dep_ind, fe_dep_ind
+  integer :: nox_flux_ind, nhy_flux_ind, atmpress_ind, xco2_ind, xco2_alt_ind
+end type MARBL_tracers_CS
+
+!> If we can post data column by column, all we need are integer
+!! arrays for ids
+! integer, allocatable, private :: id_surface_flux_diags(:) !, id_interior_tendency_diags(:)
+
+!> Module parameters
+  real, parameter :: cm_per_m = 100.
+  real, parameter :: atm_per_Pa = 1./101325.
+
+contains
+
+!> This subroutine is used to read marbl_in, configure MARBL accordingly, and then
+!! call MARBL's initialization routine
+subroutine configure_MARBL_tracers(GV, param_file, CS)
+  type(verticalGrid_type),    intent(in) :: GV   !< The ocean's vertical grid structure
+  type(param_file_type),      intent(in) :: param_file !< A structure to parse for run-time parameters
+  type(MARBL_tracers_CS),     pointer    :: CS   !< A pointer that is set to point to the control
+                                                 !! structure for this module
+
+#include "version_variable.h"
+  character(len=40)  :: mdl = "MARBL_tracers" ! This module's name.
+  character(len=256) :: log_message
+  character(len=256) :: marbl_in_line(1)
+  integer :: m, nz, marbl_settings_in, read_error
+  nz = GV%ke
+  marbl_settings_in = 615
+
+  ! (1) Read all relevant parameters and write them to the model log.
+  call log_version(param_file, mdl, version, "")
+  call get_param(param_file, mdl, "marbl_settings_file", CS%marbl_settings_file, &
+                 "The name of a file from which to read the run-time "//&
+                 "settings for MARBL.", default="marbl_in")
+  call get_param(param_file, mdl, "atm_co2_const", CS%atm_co2_const, &
+                 "Value to send to MARBL as xco2", default=284.317)
+  call get_param(param_file, mdl, "atm_alt_co2_const", CS%atm_alt_co2_const, &
+                 "Value to send to MARBL as xco2_alt_co2", default=284.317)
+
+  ! (2) Read marbl settings file and call put_setting()
+
+  ! (2a) only master task opens file
+  if (is_root_PE()) &
+     ! read the marbl_in into buffer
+     open(unit=marbl_settings_in, file=CS%marbl_settings_file, iostat=read_error)
+  call broadcast(read_error, root_PE())
```

Comment:
No need to do the broadcast here. The `MOM_error` call should bring down the entire model if the root PE fails during read.

Reply:
So the `if read_error` block should be nested in the `if is_root_PE` block? And if the read fails, it looks like any other tasks that got ahead of the root task will be waiting for the broadcast at 152 when the abort happens.